File: Bootstrap_new/a1_bootstrap.m
function a1_bootstrap % Solving the system, check
close all
clearvars
clc
format short g, format compact
profile on
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
addpath('D:\Dropbox\Naresh_Sahu\Post Monsoon ANN\Codes_today')

nB=100; %% number of Boot samples

load b0_X_GQ.mat GQ X0

load Fitlm_model.mat Modells_flm
load fitrnet_bayesian.mat Modells_ann
load random_forest.mat Modells_rf Opttable_rf
Mdl={Modells_flm Modells_ann Modells_rf};


nResp=numel(Mdl);


Boot_main=cell(1,nResp);


for j= 1:size(Mdl,2)
    XX=X0{:,:};
    YY=GQ;

    nc=size(XX,1);
    ns=size(XX,1);

    YB_sample=zeros(ns,nB);
    RMSE_Boots=zeros(1,nB);


    rng("default")
    for i=1:nB
        idx = randsample(1:nc, round(0.9*nc));%% (nc,nc,true) is with replacement
        Xdata=XX(idx,:);
        Ydata=YY(idx,:);
        boot_model=createmdl(j,Xdata,Ydata,Mdl,Opttable_rf);
        RMSE=find_RMSE(boot_model,Xdata,Ydata);
        RMSE_Boots(1,i)=RMSE;
        Yb_sample=predict(boot_model,XX);
        YB_sample(:,i)=Yb_sample;
    end
   
    Boot_main(1,j)={YB_sample};

end

save Bootconf.mat Boot_main %boot_RMSE
function RMSE=find_RMSE(mdl,X,Y)
pred=predict(mdl,X);
AE=abs(Y-pred);
SE=AE.^2;
MSE=mean(SE,1,"default");
RMSE=MSE.^0.5;

function mdl=createmdl(j,Xdata,Ydata,Mdl,Opttable_rf)
if j==1
    mdl = fitlm(Xdata, Ydata,Mdl{j}.Formula); % Fit a linear model for the first case
end
if j==2
    mdl=fitrnet(Xdata,Ydata,'InitialStepSize','auto','Activations',Mdl{j}.Activations,'Standardize',1,'LayerSizes',Mdl{j}.LayerSizes);
end
if j==3
    para1={'NumVariablesToSample',Opttable_rf.NumVariablesToSample,'MinLeafSize',Opttable_rf.MinLeafSize,'MaxNumSplits',Opttable_rf.MaxNumSplits};
    para2={'Method',Opttable_rf.Method,'NumLearningCycles',Opttable_rf.NumLearningCycles};
    para3={'PredictorSelection','interaction-curvature','Reproducible',true,'Surrogate','on','Type','regression'};
    rng(1)
    t = templateTree(para1{:},para3{:});
    rng(1)
    mdl=fitrensemble(Xdata,Ydata, 'Learners',t,para2{:});
end

File: python_code/Bootstrap_new/a1_bootstrap.py
import numpy as np
import os
import joblib
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import BaggingRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.neural_network import MLPRegressor


def find_RMSE(mdl, X, Y):
    pred = mdl.predict(X)
    AE = np.abs(Y - pred)
    SE = AE ** 2
    MSE = np.mean(SE)
    RMSE = MSE ** 0.5
    return RMSE


def createmdl(j, Xdata, Ydata, Mdl, Opttable_rf):
    if j == 0:
        mdl = LinearRegression()
        mdl.fit(Xdata, Ydata)
    elif j == 1:
        layer_sizes = Mdl['LayerSizes']
        activation = Mdl['Activations']
        act_map = {'tanh': 'tanh', 'sigmoid': 'logistic', 'relu': 'relu'}
        mdl = MLPRegressor(
            hidden_layer_sizes=layer_sizes,
            activation=act_map.get(activation, 'relu'),
            max_iter=2000,
            random_state=1,
            early_stopping=True
        )
        mdl.fit(Xdata, Ydata)
    elif j == 2:
        base_tree = DecisionTreeRegressor(
            min_samples_leaf=Opttable_rf.get('MinLeafSize', 5),
            max_depth=Opttable_rf.get('MaxNumSplits', 10),
            max_features=Opttable_rf.get('NumVariablesToSample', 6),
            random_state=1
        )
        mdl = BaggingRegressor(
            estimator=base_tree,
            n_estimators=Opttable_rf.get('NumLearningCycles', 50),
            random_state=1,
            bootstrap=True
        )
        mdl.fit(Xdata, Ydata)
    return mdl


def a1_bootstrap(GQ, X0, Modells_flm, Modells_ann, Modells_rf, Opttable_rf):
    nB = 100
    nResp = 3
    Mdl = [Modells_flm, Modells_ann, Modells_rf]
    Boot_main = [None] * nResp

    for j in range(nResp):
        XX = X0
        YY = GQ

        nc = XX.shape[0]
        ns = XX.shape[0]

        YB_sample = np.zeros((ns, nB))
        RMSE_Boots = np.zeros(nB)

        rng = np.random.default_rng(42)
        for i in range(nB):
            idx = rng.choice(nc, size=round(0.9 * nc), replace=False)
            Xdata = XX[idx, :]
            Ydata = YY[idx]
            boot_model = createmdl(j, Xdata, Ydata, Mdl[j], Opttable_rf)
            RMSE = find_RMSE(boot_model, Xdata, Ydata)
            RMSE_Boots[i] = RMSE
            Yb_sample = boot_model.predict(XX)
            YB_sample[:, i] = Yb_sample

        Boot_main[j] = YB_sample

    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    save_dir = os.path.join(base_dir, 'saved_models')
    os.makedirs(save_dir, exist_ok=True)
    np.savez(os.path.join(save_dir, 'Bootconf.npz'),
             Boot_main=np.array(Boot_main, dtype=object))
    return Boot_main